### PR TITLE
Shelley: avoid expensive equality check of the LedgerView

### PR DESCRIPTION
**Pull request by mrBliss**
We check whether the `LedgerView` changed after (re)applying each block, to see
whether we should store a snapshot of the old one. The `LedgerView` includes the
overlay schedule (a map with 432,000 entries!) and the pool distribution (a map
with an entry for each stakepool). Comparing these two data structures for
equality is very costly. Since the overlay schedule and the pool distribution
only change when transitioning to a new epoch, we can use the epoch number as a
proxy for the equality check of those two data structures.

The other two data structures, i.e., `lvProtParams` and `lvGenDelegs`, are small
and can be checked time.

### CODE REVIEW

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs [hunk 1]
```diff
@@ -279,6 +277,20 @@ applyHelper f cfg blk
...
+    ledgerViewChanged :: SL.ShelleyState c -> SL.ShelleyState c -> Bool
+    ledgerViewChanged old new
+      | SL.nesEL old /= SL.nesEL new
+        -- When the epoch changed, we know the overlay schedule and the pool
+        -- distribution will have changed. This allows us to avoid checking
+        -- the equality of these huge data structures.
+      = True
+      | otherwise
+      = SL.lvProtParams oldLedgerView /= SL.lvProtParams newLedgerView ||
+        SL.lvGenDelegs  oldLedgerView /= SL.lvGenDelegs  newLedgerView
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs:289]: @nc6 Do these two values ever change within an epoch or also only at the epoch boundary?

If the latter, then we can improve/simplify things even further.

Comment (nc6) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs:289]: Protocol parameters only change at an epoch boundary. Genesis delegates may change within an epoch.

Comment (mrBliss) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs]: Great, updated. Thanks!

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**